Submariner/SBDatabaseController.m

```diff
@@ -1364,6 +1378,27 @@ - (void)subsonicConnectionSucceeded:(NSNotification *)notification {
 
 
 
+#pragma mark -
+#pragma mark Window Notification (Private)
+
+- (void)windowDidChangeOcclusionState:(NSNotification *)notification {
+    NSWindow *sender = [notification object];
+    if ([sender isEqual:self.window]) {
+        BOOL visible = self.window.occlusionState & NSWindowOcclusionStateVisible;
+        if (visible) {
```

Comment:
I think this would install the timer even if it's stopped/paused when it becomes visible (et vice versa)

Reply:
Though I see that `updateProgress` will try to return early to avoid doing work. That does still make the timer wake up though, but if you're already playing, I guess it doesn't matter too much

Reply:
Yeah, the return early part is to prevent high CPU usage under this scenario: play a song->minimize the window->current song finished, next song started while minimized (update timer will still be installed and start running)

Reply:
Come to think about it, maybe we should *only* do the early return here and avoid all the timer juggling? Early return reduces the CPU usage basically as much as uninstall timer. Though in the pursuit of perfection, no time should be running, that I agree.

Reply:
You could check for playback before (un)installing the timer, but another way to do is is centralize the checks into early returns into those methods, so the caller doesn't need to be aware of the taste.

Reply:
You don't like the central check here inside `updateProgress`? I think it kinda beats early checks in all the playback-related methods. Under the previous scenario I mentioned, only the check inside `updateProgress` prevents the unnecessary UI refreshing.

Reply:
Good idea, but it might be better to put those checks in the timer installation (or both) and avoid having the timer tick in the first place. I'm fine either way, but it might be worth experimenting to see the impact of a timer that returns early  versus uninstalling it.

Reply:
As far as I can tell, I see no obvious difference between early return in `updateProgress` and uninstalling the timer. The CPU usage are both quite low, at around 1%-2%.

Reply:
In that case it's probably fine then.